routing/router.go

```diff
@@ -2293,3 +2293,265 @@ func generateBandwidthHints(sourceNode *channeldb.LightningNode,
 
 	return bandwidthHints, nil
 }
+
+var (
+	errDisconnectedSequence = errors.New("disconnected channel sequence")
+
+	errUnknownPolicy = errors.New("unknown channel policy")
+
+	errIncompatiblePolicies = errors.New("incompatible channel policies")
+)
+
+// BuildRoute returns a fully specified route based on a list of channels.
+func (r *ChannelRouter) BuildRoute(source route.Vertex, amt *lnwire.MilliSatoshi,
+	addresses []interface{}, finalCltvDelta int32, useMinAmt bool) (
+	*route.Route, error) {
+
+	log.Debugf("BuildRoute called: hopsCount=%v, amt=%v, useMinAmt=%v",
+		len(addresses), amt, useMinAmt)
+
+	// In the first pass, build a complete list of route pubkeys and fetch
+	// policies for hops that are specified by channel id.
+
+	// Initialize the route pubkeys list with the source node.
+	pubkeys := []route.Vertex{source}
+
+	// The channelAddress struct contains edge information that we need to
+	// verify channel constraints in the second pass.
+	type channelAddress struct {
+		policy   *channeldb.ChannelEdgePolicy
+		capacity btcutil.Amount
+	}
+
+	chanAddrs := []*channelAddress{}
+	for i, addr := range addresses {
+		var pubkey route.Vertex
+		var channelAddr *channelAddress
+		switch a := addr.(type) {
+
+		// If a channel id is specified, fetch the channel policy and
+		// endpoints. Use the endpoints to determine the next hop
+		// pubkey.
+		case uint64:
+			info, policy1, policy2, err := r.cfg.Graph.FetchChannelEdgesByID(a)
+			if err != nil {
+				return nil, fmt.Errorf("unknown channel %v: %v", a, err)
+			}
+			from := pubkeys[i]
+			var policy *channeldb.ChannelEdgePolicy
+			switch {
+			case info.NodeKey1Bytes == from:
+				policy = policy1
+				pubkey = info.NodeKey2Bytes
+			case info.NodeKey2Bytes == from:
+				policy = policy2
+				pubkey = info.NodeKey1Bytes
+			default:
+				return nil, errDisconnectedSequence
+			}
+
+			if policy == nil {
+				return nil, errUnknownPolicy
+			}
+			channelAddr = &channelAddress{
+				policy:   policy,
+				capacity: info.Capacity,
+			}
+
+		// If a pubkey is specified, just store that pubkey. Selecting a
+		// matching channel is done in the second pass.
+		case route.Vertex:
+			pubkey = a
+		default:
+			return nil, fmt.Errorf("unknown hop address type %T",
+				addr)
+		}
+
+		pubkeys = append(pubkeys, pubkey)
+		chanAddrs = append(chanAddrs, channelAddr)
+	}
+
+	// In the second pass, select channels for hops that are specified by
+	// pubkey and verify channel policies.
+	edges := make([]*channeldb.ChannelEdgePolicy, len(addresses))
+
+	// Keep running amounts for the payment amount and the min and max for
+	// this route.
+	var (
+		amtToSend    = amt
+		minAmtToSend = lnwire.MilliSatoshi(0)
+		maxAmtToSend = lnwire.MilliSatoshi(^uint64(0))
```

Comment:
I'm wondering whether it could be useful to also expose the max amt on the RPC?

Reply:
What use case do you have in mind? I'd think that the max mostly fails unless all channels are fully unbalanced.

Reply:
True, it would be mostly for curiosity and can be estimated with just min(capacities) anyway.